### PR TITLE
Fast up to date check for .NET Core projects (revised)

### PR DESCRIPTION
**Pull request by panopticoncentral**
**Risk**

Fast up to date checks carry a level of risk because if the check is wrong, we will either build when we didn't need to, or won't build when we need to. The former situation isn't any worse than what we have now. The latter situation is bad but can be mitigated in several ways:

* The user can elect to perform a Rebuild.
* We are adding a VS option to disable .NET Core project fast up to date checks entirely. (VS Option setting UI coming in a separate PR today.)
* The project can be marked with a `DisableFastUpToDateCheck` property that will disable the check for that project
* The user can add `UpToDateCheckInput` and `UpToDateCheckOutput` items to their project to work around missing checks.

This PR also adds logging to a "Project" output window that should help easily diagnose where problems are.

**Performance impact**

Build performance should improve, especially for solutions with many .NET Core projects.

**Is this a regression from a previous update?**

No

### CODE REVIEW

**Comment by panopticoncentral** [comment]
This is a significant reworking of the check to only use dataflow to track the items we need to do a fast up to date check. It's my first work with the dataflow, so let me know if something looks off or wonky.

**Comment by srivatsn** [comment]
@dotnet/project-system @lifengl @jviau as well.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/BuildUpToDateCheck.cs [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+                _configuredProject.Services.ProjectSubscription.ImportTreeSource.SourceBlock.SyncLinkOptions(),
+                _configuredProject.Services.ProjectSubscription.SourceItemsRuleSource.SourceBlock.SyncLinkOptions(new StandardRuleDataflowLinkOptions { RuleNames = ItemSchemas }),
+                _configuredProject.Services.OutputGroups.SourceBlock.SyncLinkOptions(),
+                target: new ActionBlock<IProjectVersionedValue<Tuple<IProjectSubscriptionUpdate, IProjectImportTreeSnapshot, IProjectSubscriptionUpdate, IImmutableDictionary<string, IOutputGroup>>>>(e => OnChanged(e)),
+                linkOptions: new DataflowLinkOptions { PropagateCompletion = true });
+        }
+
+        private void OnProjectChanged(IProjectSubscriptionUpdate e)
+        {
+            if (e.ProjectChanges.TryGetValue(ConfigurationGeneral.SchemaName, out var projectChange))
```
Comment (srivatsn) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/BuildUpToDateCheck.cs]: We have a GetPropertyOrDefault extension method that'll make this easier to read.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/Compile.xaml [hunk 2]
```diff
@@ -38,6 +38,12 @@
       DisplayName="Custom Tool Namespace"
       Description="The namespace into which the output of the custom tool is placed." />
 
+    <StringProperty Name="FullPath" Visible="false">
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/Compile.xaml]: What does this represent? Is this relative to the project path? If so I should use this for language service.

**Comment by panopticoncentral** [comment]
@jviau @srivatsn Responded to comments.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/Compile.xaml (hunk 2 above)
Comment (panopticoncentral) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/Compile.xaml]: It represents the fully qualified path to the item, as used in the other items like Content...

**Pull request merged**